torch/optim/swa_utils.py

```diff
@@ -7,7 +7,55 @@
 from torch.nn import Module
 from torch.optim.lr_scheduler import LRScheduler
 
-__all__ = ['AveragedModel', 'update_bn', 'SWALR']
+__all__ = [
+    'AveragedModel',
+    'update_bn',
+    'SWALR',
+    'get_ema_multi_avg_fn',
+    'get_swa_multi_avg_fn',
+    'get_ema_avg_fn',
+    'get_swa_avg_fn'
+]
+
+from torch.utils._foreach_utils import _group_tensors_by_device_and_dtype
+
+
+def get_ema_multi_avg_fn(decay=0.999):
+    @torch.no_grad()
+    def ema_update(ema_param_list, current_param_list, _):
+        # foreach lerp only handles float and complex
+        if torch.is_floating_point(ema_param_list[0]) or torch.is_complex(ema_param_list[0]):
+            torch._foreach_lerp_(ema_param_list, current_param_list, 1 - decay)
+        else:
+            for p_ema, p_model in zip(ema_param_list, current_param_list):
+                p_ema.copy_(p_ema * decay + p_model * (1 - decay))
+
+    return ema_update
+
+
+def get_swa_multi_avg_fn():
+    @torch.no_grad()
+    def swa_update(averaged_param_list, current_param_list, num_averaged):
+        diffs = torch._foreach_sub(current_param_list, averaged_param_list)
+        torch._foreach_addcdiv_(averaged_param_list, diffs, [num_averaged + 1] * len(averaged_param_list))
+
+    return swa_update
+
+
+def get_ema_avg_fn(decay=0.999):
+    @torch.no_grad()
+    def ema_update(ema_param, current_param, num_averaged):
+        return decay * ema_param + (1 - decay) * current_param
+
+    return ema_update
+
+
+def get_swa_avg_fn():
+    @torch.no_grad()
+    def swa_update(averaged_param, current_param, num_averaged):
+        return averaged_param + (current_param - averaged_param) / (num_averaged + 1)
+
+    return swa_update
 
 
 class AveragedModel(Module):
```

Comment:
Do you want to also update the documentation here to include EMA in addition to SWA?

Reply:
Done